### PR TITLE
obs-browser: Enable Linux support

### PR DESCRIPTION
**Pull request by cg2121**
This requires CEF >= 3683

### CODE REVIEW

**Comment by cg2121** [comment]
I found out with JIm's qt-message-pump branch the performance is bad. I changed it to use @bazukas suggestion in https://github.com/obsproject/obs-browser/issues/141 to use QT_NO_GLIB, and the performance is much better.

**Comment by cg2121** [comment]
Linux support still needs some more work. I cant get the browser source to work when I install it to the system, but when I install it in portable mode it works just fine.

**Comment by cg2121** [comment]
I have now reopened this because with https://github.com/obsproject/obs-browser/pull/166 everything is working as expected.

**Comment by ifohancroft** [comment]
In which OBS Studio version can we expect to see this?

**Comment by dodgepong** [comment]
This won't be in v24, which is set to come out in a few weeks, but will likely be strongly considered for v25 toward the end of the year.

**Comment by jp9000** [comment]
Probably whatever minor or major version is after 24.0.  So it could be 24.1, or 25.0 depending on whichever comes first.  I just want to make sure I can more thoroughly test it, and see if browser panels are working or if I can get them working.  I don't want to delay 24.0 further though so I want to get 24.0 first.

**Comment by WizardCM** [comment]
Alright, as v24 is out, it's time to try and get some heavy testing for this PR. Below are things to test. I'll start with the absolute basics, then build up from there.

As the current implementation is starting off without browser panels, I'll cover everything I can for browser sources first.

1. Add a browser source via the Add menu
2. Once the properties window is loaded, the default page should load automatically too, pointed to https://obsproject.com/browser-source (it may take a moment)
3. By default, width and height should be 800x600, there should be some custom CSS already defined (though it won't work on this default page), and no checkboxes in the properties should be enabled
4. Scroll down the list of properties, and click "Refresh cache of current page". The webpage should reload.
6. Check the Audio Mixer in the main window - the Browser should not currently have an entry. In the properties for this browser source, check "Control audio via OBS", then click OK. There should now be an entry.
7. Drag the source around the canvas. Create a new scene, then switch back. The source should continue to render correctly.
8. Open the properties for the existing browser, and change it to a webpage that contains audio & video. YouTube is recommended, but any site will do. Once you've changed the URL, click OK.
9. The video should play, and the mixer should move with the audio coming in. If you cannot hear the audio, go to Edit -> Advanced Audio Properties, make sure it's not set to "Monitor Off" for the browser, and also check Settings -> Audio to ensure the Monitoring Device is set either to Default, or to your main audio output device.
10. Adjust the volume in the mixer, or try hitting the Mute button. The audio of the browser should be affected.
11. Switch to another scene. The audio should mute itself. Wait a few seconds, then switch back to the scene containing the browser. The video should have skipped the few seconds you weren't watching, and the audio should play once again. Keep an eye on it, make sure the video hasn't frozen.
12. Right click the source in the list, and click "Interact". Using this window, pause the media as if it was a normal web browser. This could be a PLAY/PAUSE button, or anywhere within the video frame itself. Also try scrolling the webpage. Note that context menus currently aren't expected to function.
12. Next, let's try something with more of a performance hit. Open the properties for the source, set the resolution to 1920x1080, load http://webglsamples.org/field/field.html as the URL then click OK.
13. By default, the page will be locked to 30 FPS, same as the canvas. Validate that's the case, then go to Settings -> Video, change the frame rate to 60, and click OK. The webpage should automatically adjust to the new frame rate, and sit there consistently. You can also use the Interact screen to enable more grass. Try opening the Interact window alongside the browser's Properties window. All three (including the canvas) should be perfectly in sync, and they should maintain an FPS of 60.
14. Finally, close all extra windows, then restart OBS. Once OBS is open again, the browser should load the page automatically, and the canvas frame rate. If everything worked, delete the browser.

Now, go to Settings -> Advanced, and at the bottom uncheck "Enable Browser Source Hardware Acceleration". Restart OBS, then go through the above 14 steps again. Note that performance will be significantly worse, especially for the WebGL grass demo, but everything else should work the same.

That covers the basic and performance-heavy capabilities. Below are some edge cases:
* Try adding a follow Alert Box with both a sound effect and text to speech from https://streamlabs.com - every time the alert plays, both sounds should play (should match the test URL in your normal browser)
* Create an overlay at https://streamelements.com - especially one with webm-powered alerts. The animations should work correctly, as should sounds.
* Try each of the settings available in the properties of the source, including custom CSS. Note that some webpages don't allow custom CSS, and you'll see an error in the console "because it violates the following Content Security Policy directive"


I'll add to this comment if I come up with any other edge cases worth testing.

**Comment by ifohancroft** [comment]
@WizardCM I'll test those right now. Is the patch applied in the main branch or do I apply it myself/build from another branch?

**Comment by WizardCM** [comment]
@ifohancroft Good point, I've added instructions to the top of the comment.

**Comment by ifohancroft** [comment]
@WizardCM Thanks! I've also just realized I'm dumb and obviously it's not merged and I need to download the pull request as a patch and apply it myself to the code before building it. The question of whether I should build CEF myself still stands though so I will check the top comment for your instructions.

File: CI/install-dependencies-linux-ubuntu16.sh [hunk 1]
```diff
@@ -55,5 +55,14 @@ cmake -DENABLE_TESTING=Off -DUSE_SHARED_MBEDTLS_LIBRARY=On ../mbedtls-mbedtls-2.
 make -j 12
 sudo make install
 
+# build cef
+cd ~/projects
+curl -L "http://opensource.spotify.com/cefbuilds/cef_binary_76.1.5%2Bgd8a577c%2Bchromium-76.0.3809.87_linux64_minimal.tar.bz2" -o "cef.tar.bz2"
```
Comment (WizardCM) [inline CI/install-dependencies-linux-ubuntu16.sh]: We haven't tested our end with 3809. We're likely downgrading from 3770 (which Windows has in v24) to **3729**, so I'd use that.

Comment (Eein) [reply CI/install-dependencies-linux-ubuntu16.sh]: for cherry picking in the changes on bazukas branch here is what i did
```
cd ./plugins/obs-browser
git remote add bazuka https://github.com/bazukas/obs-browser.git
git fetch bazuka
git cherry-pick f536b1457009607a47a15bf02ef14010b647cf81
```

**Comment by Eein** [comment]
> @WizardCM Thanks! I've also just realized I'm dumb and obviously it's not merged and I need to download the pull request as a patch and apply it myself to the code before building it. The question of whether I should build CEF myself still stands though so I will check the top comment for your instructions.

for cherry picking in the changes on bazukas branch here is what i did
```
cd ./plugins/obs-browser
git remote add bazuka https://github.com/bazukas/obs-browser.git
git fetch bazuka
git cherry-pick f536b1457009607a47a15bf02ef14010b647cf81
```

**Comment by ifohancroft** [comment]
@WizardCM I have run into an error building OBS which I haven't ran into before and that isn't the error when bazuka's browser source doesn't work with a CEF version so I suppose it may have been because of something I've done as my git knowledge is very little so I have probably not followed the instructions correctly.

Here's what I did:
1. I've downloaded the regular/master branch of obs-studio by going to https://github.com/obsproject/obs-studio and copying the url, then executing via terminal: git clone https://github.com/obsproject/obs-studio.git
2. I've then checked https://github.com/obsproject/obs-studio/pull/1759/files for the changes that need to be made and since I couldn't easily export the pull request as a diff or patch file I've made them manually (it helped doing it manually as in UI/obs-app.cpp or maybe it was in plugins/CMakeLists.txt some of the functions/lines needed changing were at different line numbers so I found the appropriate functions needing changing)
3. cd'd into the obs-studio source folder/plugins/obs-browser
4. git clone'd bazuka's repo contents here
5. I've compiled successfully portable OBS following the instructions here: https://github.com/obsproject/obs-studio/wiki/Install-Instructions#linux-portable-mode-all-distros
6. I saw there is no browser source and realized the dependencies shell script isn't running automatically
7. I have deleted the compiled OBS
8. Downloaded and built CEF myself
9. Reran OBS's cmake line adding the -DBUILD_BROWSER=ON and -DCEF_ROOT_DIR="/path/to/built/cef/dir" and started compiling.
10. Got an error about missing data/locale in the main obs source folder (realized there's not even data folder there).
11. Created the data folder there and copied inside it the locale folder from my system-wide OBS Studio installation
12. Reran the cmake line again and it finished successfully.
13. Ran the make -j4 && make install line and it failed with errors:
```
/home/ifohancroft/Downloads/obs-studio/UI/auth-oauth.cpp:19:10: fatal error: browser-panel.hpp: No such file or directory
   19 | #include <browser-panel.hpp>
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [UI/CMakeFiles/obs.dir/build.make:199: UI/CMakeFiles/obs.dir/auth-oauth.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```
14. Then it seems compilation continued...
Then I got:
```
make[1]: *** [CMakeFiles/Makefile2:1419: UI/CMakeFiles/obs.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

Then continued until: [ 77%] Built target decklink-ouput-ui
Then it failed:
make: *** [Makefile:152: all] Error 2

If it is something I did wrong in the process, specially if it's related to git, please, if possible explain it to me like I am an idiot. The only git things I can do are like git clone, git status, git commit, git diff, git push

P.S. At first I got confused and built CEF, then obs-linuxbrowser on its own, then put its compiled files into the OBS source directory's browser plugin dir, then realized my mistake and deleted everything and started at #1 above. Also, during the compilation process there were the usual warnings you get when compiling most software. Some were in some OBS files about using deprecated data types or something, most were in the CEF wrapper library about a function not using a parameter its wanting or getting, etc..
```

**Comment by Eein** [comment]
CPU Name: Intel(R) Core(TM) i7-7820X CPU @ 3.60GHz
CPU Speed: 1200.012MHz
Physical Cores: 8, Logical Cores: 16
Kernel Version: Linux 5.0.0-29-generic
Distribution: "Pop!_OS" "19.04"
Window System: X11.0, Vendor: The X.Org Foundation, Version: 1.20.4
NVIDIA Corporation GeForce GTX 1070/PCIe/SSE2
**OBS 23.2.1-143-gd0029ef2-modified (linux)**

> 5 Check the Audio Mixer in the main window - the Browser should not currently have an entry. In the properties for this browser source, check "Control audio via OBS", then click OK. There should now be an entry.
There isn't a Control audio via OBS option, and the browser has an entry in the audio mixer, this branch may be still on an earlier version without the option

> 7/8 Open the properties for the existing browser, and change it to a webpage that contains audio & video. YouTube is recommended, but any site will do. Once you've changed the URL, click OK.
The video should play, and the mixer should move with the audio coming in. If you cannot hear the audio, go to Edit -> Advanced Audio Properties, make sure it's not set to "Monitor Off" for the browser, and also check Settings -> Audio to ensure the Monitoring Device is set either to Default, or to your main audio output device.

Youtube source started at full fps and slowly crawled to not rendering - cpu usage low.
Audio worked as expected

Inspect mode worked great, and if you move your mouse around the inspect area, it would maintain 60fps in both the preview and the inspect area
As soon as you stop, the video crawls to a stop per the prior issue.
Audio continues while video does not (all controls work)

> Next, let's try something with more of a performance hit. Open the properties for the source, set the resolution to 1920x1080, load http://webglsamples.org/field/field.html as the URL then click OK.

Testing 12 (the webgl) it ran at around 15fps
some text was too small to see, but adding `body { font-size: 18px }` made it usable in inspect

> Now, go to Settings -> Advanced, and at the bottom uncheck "Enable Browser Source Hardware Acceleration". Restart OBS, then go through the above 14 steps again. Note that performance will be significantly worse, especially for the WebGL grass demo, but everything else should work the same.
This does not seem to exist.
Is this upstream? (along with 5?) - can rebase if required

> Try adding a follow Alert Box with both a sound effect and text to speech from https://streamlabs.com - every time the alert plays, both sounds should play (should match the test URL in your normal browser)
Streamlabs alerts ran at full framerate without audio issues

> Create an overlay at https://streamelements.com - especially one with webm-powered alerts. The animations should work correctly, as should sounds.
Streamelements alerts ran at full framerate without audio issues

>Try each of the settings available in the properties of the source, including custom CSS. Note that some webpages don't allow custom CSS, and you'll see an error in the console "because it violates the following Content Security Policy directive"
Local files worked with custom-css no problem, CSP was invoked when it should.

All other notes worked as described.

Other notes >
Had a crash on opening filters for the browser that i could not reproduce
https://gist.github.com/Eein/7c0e0c44fefb8ed1b893ae4eed07b5d7

**Comment by Eein** [comment]
> @WizardCM I have run into an error building OBS which I haven't ran into before and that isn't the error when bazuka's browser source doesn't work with a CEF version so I suppose it may have been because of something I've done as my git knowledge is very little so I have probably not followed the instructions correctly.
...snip

I also tried portable mode at first, and browser source did not show - however mine compiled and ran fine.

**Comment by WizardCM** [comment]
> 
> 
> **CEF 3770**
> Testing 12 (the webgl) it ran at around 15fps

Turns out hardware acceleration is purely Windows only, so that'll definitely ruin the WebGL test. Thanks for validating. I've updated the steps to match.

In terms of audio control not working properly, it's possible the code to show the checkbox is Windows-only, so we'll need to update that too.

CEF 3770 seems to have a bug where it won't properly update the frames without hardware acceleration. Please test again with 3729 or 3809.

**Comment by WizardCM** [comment]
@ifohancroft You want to make sure you clone OBS with all submodules (check the OBS build instructions). On top of this, you don't want to use bazuka's branch as-is, it is far behind master and therefore OBS' UI will error due to missing code.

**Comment by Eein** [comment]
**CEF 3729**
> 11 Right click the source in the list, and click "Interact". Using this window, pause the media as if it was a normal web browser. This could be a PLAY/PAUSE button, or anywhere within the video frame itself. Also try scrolling the webpage. Note that context menus currently aren't expected to function.

Interacting works, but scrolling does not

Other notes
Tested webgl anyways with the same result, a couple extra fps, but no major improvements.

Everything else works fine! Gonna play around with filters and other things to see if i can make it break :dagger:

**Comment by ifohancroft** [comment]
@WizardCM Ah yeah. Had to pull it recursive. Since I don't know what bazuka's browser source is lacking behind the main OBS branch, if I understand your instructions above, would it be enough to just do the following?:

1. Download the appropriate version of cef and build it.
2. Recursively clone OBS's branch
3. Make the changes to OBS done in the current pull request
4. Make changes to OBS's browser-source plugin according to pull request https://github.com/obsproject/obs-browser/commit/f536b1457009607a47a15bf02ef14010b647cf81
5. Compile/Build passing the neccessary cmake variables, make and then test

**Comment by Eein** [comment]
** CEF 3809 **
refreshing cache on streamlabs still breaks here

twitch and live youtube videos not supported (may be the same with other builds and live streaming platform)

Will poke around on this version to see if i can break it :)

NOTE: It seems that the default CSS overflow hidden here
```
body { background-color: rgba(0, 0, 0, 0); margin: 0px auto; overflow: hidden; }```
caused the issue with scrolling - its working fine when removing it. 
would have to rebuild the other version of but this is likely the issue - will revise my above notes

**Comment by ifohancroft** [comment]
@WizardCM I've just tested pretty much everything you've listed in the tests.
Here are the results:

1. OK
2. OK
3. OK Except the fact that 'Use custom frame rate' is checked and is uncheckable.
4. OK
5. OK (P.S. This idea idea is great (to have the browser audio controllable by OBS because it would make it easier when you are configuring your notifications and for quick adjustments during stream on long notifications)
6. OK
7. OK
8. The video plays perfectly out of the box (at the very first second the audio sort of skipped but it probably was just loading)
9. Again - works flawlessly
10. OK
11. The rest is awesome but I can't really scroll or even have a scroll on the window. Scrolling by the mouse wheel doesn't work either.
I just have a part of the page loaded and the rest is not visible unless you increase the source's resolution. I can click and everything else though (This function is fucking awesome!, I have always been using OBS on Linux so
I had no idea it was having such features on Windows (I presume it was already available there))
12. I can adjust the framerate from the browser source's properties without going to settings first. Not sure if this is supposed to happen. (By default the FPS in Video is 30 but
I could increase the source fps to more. Not sure if it actually changed as I am testing with a youtube) After adjusting the framerate to 60 in the Settings > Video. The FPS in the Browser Properties still
sits to what I set it earlier. Changing the fps in the properties does affect the playback of the video. I couldn't notice when increasing it but when lowering it, it is clear that it affects it.
Everything is perfectly in-sync though - the scene with the source, the interact window and the properties of the source
13. OK

This is maybe more of a suggestion or question about how the browser source works other than a problem but it tries to fit too much into too little it seems, at least on my website's about me page https://www.ifohancroft.com/about
it shows everything so it squeezes it to unreadable levels and my page is responsive. I am not sure as what kind of user agent it presents itself, maybe we need an adjustable zoom level?

Custom CSS works perfectly.
Refresh browser when scene becomes active works perfectly
Not sure what Shutdown source when not visible is supposed to do so not sure how to test it.

I signed in with Streamelements but I don't usually use them so I don't know how I do setup the overlay for the test case. Sorry.

Streamlabs alert box with both text to speech and sound works perfectly and is in sync. The TTS plays after the sound but maybe that's how it's usually supposed to work in Streamlabs. I haven't used TTS before.

P.S. Once I understood what changes I'm supposed to make (that I need to recursively clone the repo and that bazuka's browser is behind so I just need to change OBS's browser using the pull request) that was very easy.

Btw since those pull requests work but have probably been done quite some time ago and the OBS code has changed since then, so as it is they can't be applied because of some difference in what is on what line, should I create a pull request that fixes that, just so you have one current?

**Comment by WizardCM** [comment]
I appreciate the extensive testing and documentation from each of you. Glad to know so many aspects seem to work out of the box, and we'll work out the kinks you ran into before releasing this.

Currently panels are Windows-only, so even macOS doesn't have them. Jim has said he wants to try and get them working for the first official release with the browser on Linux, and if he ends up getting it working I'll provide steps to test that too. Feature parity is ideal, but may take some time as this stuff is largely untested in its current form.

**Comment by Eein** [comment]
Feel free to ping me on here or discord if you need another round or another distro. I might poke around with arch or manjaro later this week and report back.

I'm also very interested in the panels and will be more than glad to help with that testing. 

If i can figure out some of the contribution requirements (testing in c++ and formatters) I might be able to contribute once all the branches are merged and its less of a hassle :)

**Comment by WizardCM** [comment]
Friendly bump: this PR has merge conflicts.

**Comment by ifohancroft** [comment]
Not sure if this helps but:
@cg2121

**Comment by christophercrouzet** [comment]
> P.P.S. Off-topic but when are we expecting to get the Twitch integration/Twitch Chat in OBS on Linux and is anything I can do to help with that?

> Currently panels are Windows-only, so even macOS doesn't have them. Jim has said he wants to try and get them working for the first official release with the browser on Linux, and if he ends up getting it working I'll provide steps to test that too. Feature parity is ideal, but may take some time as this stuff is largely untested in its current form.

Here's to show some additional interest—I'm very much looking forward to having this feature made available for Linux! :v:

**Comment by WizardCM** [comment]
I have now tested this PR myself, and can confirm things generally work correctly.

**Comment by Francexi** [comment]
any idea on the time of the merging of this pull? This would be a lot helpful

**Comment by MartinPL** [comment]
Is there any Arch build of this branch online?

**Comment by Fenrirthviti** [comment]
> Is there any Arch build of this branch online?

You would need to compile yourself. We only maintain official builds for Ubuntu at the current time.

File: CI/install-dependencies-linux-ubuntu16.sh (hunk 1 above)
Comment (cg2121) [reply CI/install-dependencies-linux-ubuntu16.sh]: Its been changed to 3770 now.

**Comment by shadow-light** [comment]
Keyboard input seems to be ignored when interacting with the browser (mouse input works). Just checking whether this is already known or not?

**Comment by dx9s** [comment]
Not sure if this is well known but the native OBS Browser source is painfully slow on Linux. I have to use the (no longer maintained by developer) linux-browser source ( https://github.com/bazukas/obs-linuxbrowser ) source to get high frame rates.

A good test is to find a web page with WebGL animations on it such as.

https://makebackground.io/bk/network/blue-network-connection-graph#

On "native" browser source, I get around 2-4 frames a second a second?!

Using the linux-browser source, I get whatever FPS is set at. 30 FPS or 60 FPS is perfectly usable.

**Comment by kkartaltepe** [comment]
Merged pull requests are not the place to file bugs or ask for support. Consider asking for support in our discord or filing a bug report if you think this is a bug. Please also use the correct repository, obs-studio is not the correct repository for obs-browser bugs.
